[Transforms] Debug values are not remapped when cloning.

CarlosAlbertoEnciso opened the pull request:
When cloning instructions from one basic block to another,
the debug values are not remapped, in the same was as the
normal instructions.

CarlosAlbertoEnciso committed:
[Transforms] Debug values are not remapped when cloning.

When cloning instructions from one basic block to another,
the debug values are not remapped, in the same was as the
normal instructions.

OCHyams commented on llvm/lib/IR/IntrinsicInst.cpp:
```diff
@@ -136,6 +137,8 @@ void DbgVariableIntrinsic::replaceVariableLocationOp(Value *OldValue,
   auto Locations = location_ops();
   auto OldIt = find(Locations, OldValue);
   if (OldIt == Locations.end()) {
+    if (AllowEmpty)
+      return;
```
Why is this needed now (can you point to a test that fails without it)?

CarlosAlbertoEnciso replied:
The included test fails:
`llvm/test/Transforms/CallSiteSplitting/callsite-split-debug.ll`

This is the line that triggers the assertion:
`  call void @llvm.dbg.value(metadata !DIArgList(i16 %c, i16 %c), metadata !7, metadata !DIExpression()), !dbg !8
`

CarlosAlbertoEnciso replied:
Also, looking at the `DbgVariableRecord::replaceVariableLocationOp` the condition should be:
```
    if (AllowEmpty || DbgAssignAddrReplaced)
      return;
```

OCHyams replied:
Ah I see, yeah that makes sense. Thanks!

OCHyams reviewed (commented):
Changes all look good, with one inline question.

CarlosAlbertoEnciso committed:
[Transforms] Debug values are not remapped when cloning.

When cloning instructions from one basic block to another,
the debug values are not remapped, in the same was as the
normal instructions.

Missing condition '|| DbgAssignAddrReplaced' to account
for any 'dbg.assign' replacement.

OCHyams reviewed (approved):
LGTM

CarlosAlbertoEnciso committed:
[Transforms] Debug values are not remapped when cloning.

When cloning instructions from one basic block to another,
the debug values are not remapped, in the same was as the
normal instructions.

- Make 'RemapDebugVariable' a general function by moving
  it to 'Transform/Utils/Local'.
- Change in 'JumpThreading' parameters, return values and
  local instances:
   DenseMap<Instruction *, Value *> to ValueToValueMapTy.

CarlosAlbertoEnciso commented:
@OCHyams Thanks for your review.
Moved the `RemapDebugVariable` code to `Local` as a general function.

OCHyams commented on llvm/lib/Transforms/Scalar/JumpThreading.cpp:
```diff
@@ -2008,14 +2007,15 @@ void JumpThreadingPass::updateSSA(
-                                     BasicBlock::iterator BE, BasicBlock *NewBB,
-                                     BasicBlock *PredBB) {
+
+void JumpThreadingPass::cloneInstructions(ValueToValueMapTy &ValueMapping,
+                                          BasicBlock::iterator BI,
+                                          BasicBlock::iterator BE,
+                                          BasicBlock *NewBB,
+                                          BasicBlock *PredBB) {
```
Why is it necessary to change the return value to an out-parameter here?

CarlosAlbertoEnciso replied:
Due to the compilation error: `use of deleted function`

```
  // ValueMap can't be copied nor moved, because the callbacks store pointer to
  // it.
  ValueMap(const ValueMap &) = delete;
  ValueMap(ValueMap &&) = delete;
  ValueMap &operator=(const ValueMap &) = delete;
  ValueMap &operator=(ValueMap &&) = delete;
```
`using ValueToValueMapTy = ValueMap<const Value *, WeakTrackingVH>;`

`ValueMap` can't be copied nor moved.

OCHyams commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -3648,6 +3648,33 @@ DIExpression *llvm::getExpressionForConstant(DIBuilder &DIB, const Constant &C,
+void llvm::remapDebugVariable(ValueToValueMapTy &Mapping, Instruction *Inst) {
+  auto RemapDebugOperands = [&Mapping](auto *DV, auto Set) {
+    for (auto *Op : Set)
+      if (Instruction *InstOp = dyn_cast<Instruction>(Op)) {
```
I don't think that we need the `dyn_cast` checks now that you're passing in a `ValueToValueMapTy` given that the key is a `Value *`. Could you remove them please?

OCHyams commented on llvm/include/llvm/Transforms/Utils/Local.h:
```diff
@@ -478,6 +479,10 @@ void hoistAllInstructionsInto(BasicBlock *DomBlock, Instruction *InsertPt,
 DIExpression *getExpressionForConstant(DIBuilder &DIB, const Constant &C,
                                        Type &Ty);
 
+/// Remap the debug intrinsic instructions in the \p Mapping using the
+/// \p Inst as a criteria.
```
I think the wording of this comment is slightly unclear. What do you think of this instead?

"Remap the operands of the debug records attached to \p Inst, and the operands of \p Inst itself if it's a debug intrinsic."

OCHyams reviewed (approved):
Just two small things, then this LGTM. Thanks!

CarlosAlbertoEnciso committed:
[Transforms] Debug values are not remapped when cloning.

When cloning instructions from one basic block to another,
the debug values are not remapped, in the same was as the
normal instructions.

- Rewording a comment to add more clarity.
- Remove not required 'dyn_cast.

CarlosAlbertoEnciso commented:
@OCHyams Thanks for your review.

OCHyams commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -3648,6 +3648,33 @@ DIExpression *llvm::getExpressionForConstant(DIBuilder &DIB, const Constant &C,
+void llvm::remapDebugVariable(ValueToValueMapTy &Mapping, Instruction *Inst) {
+  auto RemapDebugOperands = [&Mapping](auto *DV, auto Set) {
+    for (auto *Op : Set)
+      if (auto *InstOp = Op) {
```
You still have the `if` statements, but they're not needed either. Please can you remove those?

OCHyams replied:
Same again below - I don't think `if (Instruction *Addr = dyn_cast<Instruction>(DA->getAddress())) {` is needed on line 3661.

CarlosAlbertoEnciso replied:
@OCHyams Reworked the patch to remove those `if`s. Thanks.

CarlosAlbertoEnciso committed:
[Transforms] Debug values are not remapped when cloning.

When cloning instructions from one basic block to another,
the debug values are not remapped, in the same was as the
normal instructions.

- After removing the 'dyn_cast' checks, the following 'if's
  are not required.

SLTozer reviewed (approved):
Comments look sufficiently addressed, LGTM

CarlosAlbertoEnciso commented:
@SLTozer Thanks.
